Extend the template for "choose email" of GitHub auth.

chdinesh1089 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
last issue in https://github.com/zulip/zulip/issues/12638

* Also includes refactors in `test_auth_backends` and addition of a new test.

**Testing Plan:** <!-- How have you tested? -->
added a test

**GIFs or Screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

chdinesh1089 commented on zproject/backends.py:
```diff
@@ -987,23 +987,34 @@ def social_associate_user_helper(backend: BaseAuth, return_data: Dict[str, Any],
         if not chosen_email:
             avatars = {}  # Dict[str, str]
-            existing_account_emails = []
+            verified_existing_account_emails = []
```
renamed this for clarity. we have `existing_account_emails` for unverified accounts also. Thought it's be better to rename.

showell commented:
@timabbott The first commit here's a no-brainer extraction.  I don't know enough about the auth system to fully review the second one.

mateuszmandera commented on zerver/tests/test_auth_backends.py:
```diff
@@ -1633,6 +1640,13 @@ def social_auth_test_finish(self, result: HttpResponse,
+                for email_data_dict in email_data:
+                    email = email_data_dict["email"]
+                    if not email.endswith("noreply.github.com"):
+                        self.assert_in_success_response([], result)
```
Should we also assert that if it does end with noreply.github.com it must not appear in the response?

chdinesh1089 replied:
It probably is a good thing to do. adding that too.

mateuszmandera commented:
Looks good! Two ideas, for separate commits I suppose:
1. Since `email_data` and `expect_choose_email_screen` mean nothing in non-github test classes, could we have them in in ``**kwargs`` instead of explicitly named? And then have them explicitly named in the github overrides of those methods?
2.filter_usable_emails and get_unverified_emails are super analogical, perhaps we could refactor in such a way: `filter_usable_emails` returns all emails except the @noreply.github.com ones  and then `get_unverified/verified_emails` functions do `emails = [email_obj['email'] for email_obj in self.filter_usable_emails(...) if [not] email_obj.get('verified')]`? Perhaps that'd be cleaner

chdinesh1089 commented on zproject/backends.py:
```diff
@@ -1309,17 +1311,15 @@ def get_verified_emails(self, *args: Any, **kwargs: Any) -> List[str]:
         # @noreply.github.com/@users.noreply.github.com email
         # addresses, because structurally, we only want to allow email
         # addresses that can receive emails, and those cannot.
+        emails = self.get_all_associated_email_objects(*args, **kwargs)
```
yes, Changed it.

chdinesh1089 commented on zerver/tests/test_auth_backends.py:
```diff
@@ -782,6 +781,11 @@ def social_auth_test(self, account_data_dict: Dict[str, str],
+            if "expect_choose_email_screen" in extra_data:
+                expect_choose_email_screen = extra_data["expect_choose_email_screen"]
+            else:
+                expect_choose_email_screen = False
```
Yea, It'll be much better to avoid `expect_choose_email_screen` anywhere in `SocialAuthBase`. Change it! Thanks!

mateuszmandera commented:
@chdinesh1089 I replaced the last commit with my own demonstration of what I meant, because I figured it's kind of difficult to explain in words :slightly_smiling_face: And inserted some extra cleanup as the second commit.

Oh, now that I'm looking through it, I think we forgot this https://github.com/zulip/zulip/pull/14386#discussion_r403029922?

chdinesh1089 commented:
> Oh, now that I'm looking through it, I think we forgot this #14386 (comment)?

Ohh, sorry, I made the changed and stashed it. added that too now.

> I replaced the last commit with my own demonstration of what I meant

Okay, now it's clear what you were saying. I think I can understand the benefit of explicitly having `expect_choose_email_screen` and `email_data` as args in GitHub test class. (We can avoid the if statements inside the function and also suggest the developer to give those args when adding a new test, right?)

mateuszmandera commented:
> Okay, now it's clear what you were saying. I think I can understand the benefit of explicitly having `expect_choose_email_screen` and `email_data` as args in GitHub test class. (We can avoid the if statements inside the function and also suggest the developer to give those args when adding a new test, right?)

Yes, and it should be much less confusing for someone going through Google tests not to have to look at logic that's applicable only to Github and makes no sense in Google context.

Okay, things look good to me now! @timabbott Can you review?

timabbott commented on templates/zerver/social_auth_select_email.html:
```diff
@@ -57,8 +57,26 @@ <h1 class="get-started">Select account</h1>
+        If you'd like to use one of these to login to Zulip via GitHub,<br>
+        you must <a href="https://github.com/settings/emails">verify it.</a>
     </p>
+    {% endif %}
```
After playing with this design that shows the actual emails, it gives way too much weight to what is potentially basically useless data.  So I removed this, limiting it to just showing output if there are any.  That should be good enough for the use case; they can see the list in GitHub after clicking the link.

timabbott commented on zerver/tests/test_auth_backends.py:
```diff
@@ -1627,6 +1610,42 @@ class GitHubAuthBackendTest(SocialAuthBase):
+        result = self.client_get(self.AUTH_FINISH_URL,
+                                 dict(state=csrf_state), **headers)
+        if email_data is None:
+            email_data = [account_data_dict]
```
This `email_data` hack struck me as wrong; and I confirmed it was wrong when I added a bit more validation below.

The updated test uses `self.email_data`, which does what we need here and allows the more extensive validation below to be correct.

timabbott commented on zerver/tests/test_auth_backends.py:
```diff
@@ -1627,6 +1610,42 @@ class GitHubAuthBackendTest(SocialAuthBase):
+    def social_auth_test_finish(self, result: HttpResponse,
+                                account_data_dict: Dict[str, str],
+                                headers: Dict[str, Any]={},
+                                expect_choose_email_screen: bool=False,
```
I think some of the `headers` / `expect_choose_email_screen` refactoring refactoring is counterproductive, since it obscures what's happening and squashies together unrelated objects.  I undid that before merging.

timabbott commented:
Merged as the series ending with a3329f288c196d7c371ba5b70a09149020885a91, see the comments above on some changes (others are clear from reading the code).  The result is pretty clean and simple.  

Thanks @chdinesh1089 and @mateuszmandera!
